[P036] Memory optimization (issue #2799) and other improvements

uwekaditz opened the pull request:
CHG: Memory optimization (issue #2799)
CHG: Added setting for 'Disable scrolling while WiFi is not connected',
         scrolling is not smooth as long as the ESP tries to connect Wifi (Load 100%!)
BUG: Wifi-Bars not fully right for 128 pixel wide displays
BUG: clearing window for scrolling lines was too high and to narrow
BUG: content of lines with pixel width > 255 had to be shortened
CHG: scrolling lines improved
CHG: variables renamed for better understanding

Fixes: #2312
Closes: #3026 
Fixes: #3127 
Fixes: #2760 
Fixes: #2799 
Fixes: #2220
Fixes: #2118
Fixes: #2085
Fixes: #799


Fixes: #1377  (Should be tested, but in theory it should now be possible to have multiple instances of this plugin)

uwekaditz committed:
merged Release mega-20200328

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

uwekaditz committed:
Merge remote-tracking branch 'upstream/mega' into UweKaditz

uwekaditz committed:
Merge tag 'mega-20200426' into UweKaditz

uwekaditz committed:
[P036] Memory optimization (issue #2799)

CHG: Memory optimization (issue #2799)
CHG: Added setting for 'Disable scrolling while WiFi is not connected', scrolling is not smooth as long as the ESP tries to connect Wifi (Load 100%!)
BUG: Wifi-Bars not fully right for 128 pixel wide displays
BUG: clearing window for scrolling lines was too high and to narrow
BUG: content of lines with pixel width > 255 had to be shortened
CHG: scrolling lines improved
CHG: variables renamed for better understanding

uwekaditz committed:
Merge branch 'mega' into UweKaditz

uzi18 commented on src/_P036_FrameOLED.ino:
```diff
@@ -409,9 +511,12 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
+        P036_setBitToUL(lSettings, 16, isFormItemChecked(F("p036_pin3invers")));                // Bit 16 Pin3Invers
+        P036_setBitToUL(lSettings, 17, isFormItemChecked(F("p036_ScrollLines")));               // Bit 17 ScrollLines
+        P036_setBitToUL(lSettings, 18, !isFormItemChecked(F("p036_NoDisplay")));                // Bit 18 NoDisplayOnReceivingText
+        P036_setBitToUL(lSettings, 19, !isFormItemChecked(F("p036_ScrollWithoutWifi")));        // Bit 18 ScrollWithoutWifi
```
bit 19

uwekaditz replied:
The comment has been changed in the new version.

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -297,6 +375,11 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
+#ifdef PLUGIN_036_DEBUG
+          addLog(LOG_LEVEL_INFO, F("P036_PLUGIN_WEBFORM_LOAD ..."));
+#endif  // PLUGIN_036_DEBUG
+        P036_data_temp = new P036_data_struct();  // just temporary for Plugin_036_loadDisplayLines()
```
There is a number of times you do a `new` on this pointer, but I do miss a few calls to delete.

Are you sure this is not causing a memory leak, every time you look at the web page for this plugin?

uwekaditz replied:
In the new version there is a Delete.
I thought it will be automatically deleted while leaving the function.

TD-er replied:
If you want that kind of behavior you have to use one of those smart pointer wrappers to store the pointer.
That's always a more safe way to store data and we could use more of them in a lot of places in the code.
But as long as you don't use one of those, you have to delete them manually.

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -391,12 +480,25 @@ boolean Plugin_036(uint8_t function, struct EventStruct *event, String& string)
+          addLog(LOG_LEVEL_INFO, F("P036_PLUGIN_WEBFORM_SAVE ..."));
+#endif  // PLUGIN_036_DEBUG
+          P036_data_temp = P036_data;             // updating initialized values
+        }
```
This section I don't understand.
If the `P036_data` is not available, then you create one... that part I do (somewhat) understand.
But why do you copy the pointer to P036_data_temp?

This makes the administration of the life time of the data object next to impossible.
If there is only a single owner of the pointer, then you know you have to delete the object before re-assigning a new one (or setting it to `nullptr`)
But if there are other pointers out there pointing to the same object, then you can't be sure if someone is de-referencing it after the object has been destroyed.
Or it is never deleted and then we're having a memory leak.  (or 2 versions of the same data)

uwekaditz replied:
Changed, not using a temporary pointer in `PLUGIN_WEBFORM_SAVE` anymore.
The original idea was to update the current values if the task is already initialized.
Calling `SaveCustomTaskSettings()` was coded with the pointer -> changed.

uwekaditz committed:
[P036] temporary pointer to P036_data_struct was not deleted

BUG: temporary pointer to P036_data_struct was not deleted in PLUGIN_WEBFORM_LOAD
CHG: PLUGIN_WEBFORM_SAVE does not use a temporary pointer to P036_data_struct
CHG: Comments

TD-er commented:
It looks like you are basing this PR on a relatively older version of the plugin, as some functions you seem to call have been removed about a week ago.

Also there is another pull request which conflicts with your changes: https://github.com/letscontrolit/ESPEasy/pull/3026

uwekaditz commented:
Oh I see.
My code was based on the latest release mega-20200426.
I didn't notice that tonhuisman changed the plugin by the pull request #3026 
I'm going to merge it together.

uwekaditz committed:
Merge remote-tracking branch 'upstream/mega' into UweKaditz

uwekaditz committed:
Merge branch 'UweKaditz' of https://github.com/uwekaditz/ESPEasy into UweKaditz

TD-er commented:
Maybe it is an easier merge if I merge it first with the Mega branch (Ton's PR I mean)

I was already about to merge it yesterday.

uwekaditz committed:
[P036] merged with Tonis PR #3026

CHG merged with Tonis PR #3026
CHG: Jump to any page (Display button or command) is always without page scrolling
CHG: Jump to next page (Display button) is not allowed while page scrolling is running

uwekaditz commented:
While merging with Tonis PR #3026 I found one problem:
In his code he starts a new page within the `PLUGIN_TEN_PER_SECOND `call.
Preparing the new page takes approx. 60ms!
I hope this is no problem.

uwekaditz commented:
By the way I got another question:
Is it possible to reset or restart the `schedule_task_device_timer()` within the the `PLUGIN_TEN_PER_SECOND` or `PLUGIN_WRITE `calls?
It would be fine to start the next page display exactly after the time given in `Settings.TaskDeviceTimer[event->TaskIndex]` after a manual jump to any page happened.

uzi18 commented:
you can use plugin timer as in rfid plugins

uwekaditz commented:
Thanks!

TD-er commented:
> Is it possible to reset or restart the schedule_task_device_timer() within the the PLUGIN_TEN_PER_SECOND or PLUGIN_WRITE calls?

Yes you can reschedule it at any time.
What I usually do is call this schedule function with `millis() + 10` or something similar.
This way it is not blocking for scheduled jobs that were scheduled to run right after the current job.

> In his code he starts a new page within the PLUGIN_TEN_PER_SECOND call.
> Preparing the new page takes approx. 60ms!

Well, if it is called from the PLUGIN_ONCE_PER_SECOND call, it would also take 60 msec.
So either way, it is blocking code.
Would be nice if it can be split into multiple small chunks as 60 msec can also have an effect on WiFi communication for example.

uwekaditz committed:
[P036] re-schedule Settings.TaskDeviceTimer after JumpToPage

CHG: re-schedule Settings.TaskDeviceTimer after JumpToPage to avoid any scheduled page change while jumping to pages
CHG: correct calculation of the page indicator counts

uwekaditz commented:
While compiling with the latest changes in mega I get following error in Atom
```
ImportError: No module named pygit2:
  File "C:\users\uwe\.platformio\penv\lib\site-packages\platformio\builder\main.py", line 154:
    env.SConscript(item, exports="env")
  File "C:\Users\Uwe\.platformio\packages\tool-scons\script\..\engine\SCons\Script\SConscript.py", line 541:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Users\Uwe\.platformio\packages\tool-scons\script\..\engine\SCons\Script\SConscript.py", line 250:
    exec _file_ in call_stack[-1].globals
  File "F:\GitHub\ESPEasy\tools\pio\generate-compiletime-defines.py", line 7:
    from pygit2 import Repository
```
In **_platformio.ini_** I had to change the entry
```
[extra_scripts_default]
extra_scripts             = pre:tools/pio/generate-compiletime-defines.py
                            tools/pio/copy_files.py
```
to
```
[extra_scripts_default]
extra_scripts             = tools/pio/copy_files.py
```
to get it running again.

uzi18 commented:
use command:
`pip install pygit2`
remember to use python3 or pip3 if available

uwekaditz commented:
Thanks, `pip install pygit2` did work.

uwekaditz commented:
I could install pygit2 but can still not compile in Atom.
```
pip install pygit2
Collecting pygit2
  Downloading pygit2-1.2.1-cp38-cp38-win32.whl (1.0 MB)
     |████████████████████████████████| 1.0 MB 1.7 MB/s
Requirement already satisfied: cached-property in c:\users\uwe\appdata\local\programs\python\python38-32\lib\site-packages (from pygit2) (1.5.1)
Requirement already satisfied: cffi>=1.4.0 in c:\users\uwe\appdata\local\programs\python\python38-32\lib\site-packages (from pygit2) (1.14.0)
Requirement already satisfied: pycparser in c:\users\uwe\appdata\local\programs\python\python38-32\lib\site-packages (from cffi>=1.4.0->pygit2) (2.20)
Installing collected packages: pygit2
Successfully installed pygit2-1.2.1
```
The installation created two folders within `\python36-32\Lib\site-packages\`
```
pygit2-1.2.1.dist-info
pygit2
```
Inside the folders and nowhere else is a file called pygit2.py!

uzi18 commented:
Atom like VSCode uses own copy of platformio/virtualenv  did you used pip in atom terminal?

uwekaditz commented:
No, I used my python environment.
The terminal in atom is not working. It is always empty and text I key in is not shown.
![grafik](https://user-images.githubusercontent.com/55511593/81120295-b23e9400-8f2c-11ea-9200-dc5e5b0f5bd6.png)
after confirming by ENTER the terminal is still empty
![grafik](https://user-images.githubusercontent.com/55511593/81120332-c97d8180-8f2c-11ea-9492-552b78798643.png)

uzi18 commented:
In my Atom have got: menu PlatformiO->Terminal->New Terminal

uwekaditz commented:
I got the terminal in atom running. I had to uninstall and to install it again.
Nevertheless I can't install pygit2
```
  building 'pygit2._pygit2' extension
  error: Microsoft Visual C++ 9.0 is required. Get it from http://aka.ms/vcpython27
  ----------------------------------------
  Failed building wheel for pygit2
```
Do I need to install Microsoft Visual C++ 9.0? I don't want to buy a license!

uzi18 commented:
show me output of:
python --version
whereis python

uwekaditz commented:
I installed Microsoft Visual C++ Compiler for Python 2.7 and got following error:

```
  building 'pygit2._pygit2' extension
  creating build\temp.win-amd64-2.7\Release\src
  C:\Users\Uwe\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\amd64\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG "-IC:\Program Files\libgit2\include" -IC:\Users
\Uwe\.platformio\python27\include -Ic:\users\uwe\.platformio\penv\PC /Tcsrc\blob.c /Fobuild\temp.win-amd64-2.7\Release\src\blob.obj
  blob.c
  c:\temp\pip-install-czpdwk\pygit2\src\blob.h(33) : fatal error C1083: Cannot open include file: 'git2.h': No such file or directory
  error: command 'C:\\Users\\Uwe\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\amd64\\cl.exe' failed with exit status 2
  ----------------------------------------
  Failed building wheel for pygit2
```

uzi18 commented:
you need python 3.7 or 3.8

uwekaditz commented:
Atom:
PS F:\GitHub\ESPEasy> python --version
Python 2.7.13
PS F:\GitHub\ESPEasy> whereis python
Did not work! 

Windows:
I have the python 3.8.1 already installed but atom is not using it:
\AppData\Local\Programs\Python\Python38-32>python --version
Python 3.8.1

uzi18 commented:
you need to find how to update python/platformio on Atom

uwekaditz commented:
I got it, finally.
I deleted all files in %USERPROFILE%\.platformio\python27\ and copied all files from the already installed version Python38-32 into this folder.
Afterwards I could install pygit2 and compiling in Atom works.
Big deal, never thought that it would be so complicated and so late ...

uzi18 commented:
You should use apm command to upgrade python package or something like that ;)
It is better to switch to VSCode it is more reliable and working better.
Or ... use gitpod as cloud editor ;)

uwekaditz committed:
Merge branch 'mega' into UweKaditz

TD-er commented:
> Big deal, never thought that it would be so complicated and so late ...

I know your feeling.
Been there also a while ago, when I somehow managed to mess up my Python environment in Windows.
My situation back then was that some tool actively searched through the hard drive to find some Python environment and messed around with it, while it was the environment used for PlatformIO.
Also it is very well possible to get more than one Python environment for Atom + PlatformIO without ever knowing it.

I should try to document a step-by-step tutorial for it as quite a lot of people are apparently running into this.

TD-er commented:
> Do I need to install Microsoft Visual C++ 9.0? I don't want to buy a license!

For any future reference... Neither do I so if you run into these kind of questions, remember the solution where you really need to buy a license is probably not the only solution.

Doesn't mean I oppose software licensing, but just that I would not make it something that is really needed to make it all work.
After all, I do make open source software myself, so it would be really strange if there is no way to build it without a paid license of some kind.

uzi18 commented:
everyone did get in same problem with transition, but after all thanks to this i did some cleanups with virtualenvs,python2,python3 and local copy of vscode etc.

have system wide python2 and python3 now and all subpackages for pip is installed via --user option to not mess with system

uwekaditz commented:
I will do a simple step-by-step instruction for installing pygit2 an atom.

With the latest mega sources I encountered another problem but solved it by myself.
### Background:
In my house I run around 10 different ESPeasy applications. The source code for these applications differs a bit from the official mega sources

- [P001] detecting flushing inputs (rememeber?)
- simple web page just showing task names and their data (better overview on my smart phone)

That's why I have a local working directory on my hard drive, not checked out from github but using TortoiseSVN and AraxisMerge to keep it updated with my github repository.

So far I got never a problem compiling it, but with the latest mega sources strange error messages came up in the pre-compile phase and the compiling stopped.
### Solution:
I had to copy the complete .git folder into my working directory. Now it works fine again.

TD-er commented:
That's strange, as the Python environment is not stored in the .git folder.
In the .git folder the Git history and information about what branches exist and are checked out is stored. Not the Python setup.

uwekaditz commented:
The .git problem is not related to python.
It is related to the setting in the pre-compile phase. Somehow there is a search for information within the .git folder.
If the folder does not exist the pre-compile failed and the complete compilation by PlatformIO is cancelled.

TD-er commented:
Ah that may have something to do with the pygit2 check I do.
I guess it makes sense to put that in a try/catch block, if you haven't made a clone, but just downloaded the zip.

uwekaditz commented:
### Step-by-Step instruction for installing pygit2 in atom (PlatformIO) environment under Windows

- close Atom
- download the latest python version from [https://www.python.org/downloads/](url)
- install python
- locate the python folder for PlatformIO, e.g. `%USERPROFILE%\.platformio\python27\`
- delete all files and directories within the folder `python27` but not the folder itself
- locate the default installation folder of python e.g. `%USERPROFILE%\AppData\Local\Programs\Python\Python38-32\`
- copy all files and folders from this python folder into the folder `%USERPROFILE%\.platformio\python27\`
- open atom
- open a terminal in atom via PlatformIO->Terminal->New terminal
- if the terminal shows just a blank window as it was in my atom installation uninstall _**Platformio Ide Terminal**_ and re-install it afterwards (via PlatformIO->Settings->Platformio Ide Terminal
- enter `pip install pygit2` in the terminal
- the pygit2 installation will start and download all necessary files

TD-er commented:
@uwekaditz Can you also post that in this sticky topic? https://github.com/letscontrolit/ESPEasy/issues/3058

uwekaditz commented:
Done!

uwekaditz commented:
Thanks uzi18 for the advice to use vscode.
It works much better than Atom, I was always looking for an IDE tha can go the reference or definition of a function or variable.
In Atom it didn't work but in vscode it works very well.

uwekaditz committed:
[P036] clearing window for scrolling lines was 1pix too large in y direction

CHG: clearing window for scrolling lines was 1pix too large in y direction
CHG: font settings for 64x48 updated

uzi18 commented:
@uwekaditz eclipse is far better here, platformio supports eclipse project generation but ... you can use only one env ;)
you can try it

uwekaditz committed:
Merge tag 'mega-20200515' into UweKaditz

TD-er commented:
I am not sure now what to do with PR #3026 

Are all changes of PR #3026  now also included in this PR?

TD-er committed:
Merge branch 'mega' into UweKaditz

uwekaditz commented:
Of course are all changes of PR #3026 included in my new code and it works fine.
See the comments in _P036_FrameOLED.ino
```
// @uwekaditz: 2020-05-11
// CHG: clearing window for scrolling lines was 1pix too large in y direction
// CHG: font settings for 64x48 updated
// @uwekaditz: 2020-05-05
// CHG: re-schedule Settings.TaskDeviceTimer after JumpToPage to avoid any scheduled page change while jumping to pages
// CHG: correct calculation of the page indicator counts
// @uwekaditz: 2020-05-04
// BUG: temporary pointer to P036_data_struct was not deleted in PLUGIN_WEBFORM_LOAD
// CHG: PLUGIN_WEBFORM_SAVE does not use a temporary pointer to P036_data_struct
// CHG: Jump to any page (Display button or command) is always without page scrolling
// CHG: Jump to next page (Display button) is not allowed while page scrolling is running
// @uwekaditz: 2020-05-03
// CHG: Memory optimization (issue #2799)
// CHG: Added setting for 'Disable scrolling while WiFi is not connected', scrolling is not smooth as long as the ESP tries to connect Wifi (Load 100%!)
// BUG: Wifi-Bars not fully right for 128 pixel wide displays
// BUG: clearing window for scrolling lines was too high and to narrow
// BUG: content of lines with pixel width > 255 had to be shortened
// CHG: scrolling lines improved
// CHG: variables renamed for better understanding
// @tonhuisman: 2020-04-13
// CHG: Added setting for 'Step through frames with Display button', when enabled, pressing the button while the display is on
// cycles through the available frames.
// CHG: Added command oledframedcmd,frame,[0..n] to select the next frame (0) or explicitly frame 1..n, where n is the last frame available
// IMP: Improvement to immediately show the frame the text is placed on when text is received.
```

TD-er commented:
OK, then I will make sure the PR will build again and then make a test build for this one.
So @tonhuisman can also test to see if all of his work is also included/working.

Since this is one of the most used plugins, we really need to be sure it is still working like it should.

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er commented:
I've made a test build for this PR:
- [ESPEasy_ESP32_mega-20200515-151-PR_3050.zip](https://www.dropbox.com/s/yck4j52o686xliq/ESPEasy_ESP32_mega-20200515-151-PR_3050.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200515-151-PR_3050.zip](https://www.dropbox.com/s/dvzg44vsvjxyd6o/ESPEasy_ESP82xx_mega-20200515-151-PR_3050.zip?dl=0)

uwekaditz commented:
I compiled the latest Mega sources by myself.
The _P036 plugin runs fine on a 64x48 and a 128x64 OLED display.

uwekaditz committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
[Framed OLED] Add checks for display initialized

TD-er commented:
I did a quick test with this PR and it crashed continuously.

Even loading the web page for this plugin causes immediate crashes.

I added some quick checks to the code, which allows it to start the ESP again with this plugin active, but it still crashes when loading the settings.

I was using these (existing) settings:
![image](https://user-images.githubusercontent.com/3751318/85294748-5198f580-b49f-11ea-986c-d6cb80043402.png)

uwekaditz commented:
I was able to reproduce the problem.
The function display_indicator() crashes!
I commented out the function and the ESP runs fine again, even while displaying just the header because all display lines are empty.
Sending text to the display using the command `oledframedcmd,<display>,<value>` works fine, too.

uwekaditz commented:
Bug will be fixed with the next commit.

uwekaditz committed:
[P036] Fixed: display_indicator() crashed

BUG: MaxFramesToDisplay was not updated if all display lines were empty -> display_indicator() crashed due to memory overflow
CHG: MaxFramesToDisplay will be updated after receiving command with new line content
CHG: Added some checks if P036_data are valid and if P036_data->isInitialized()

TD-er commented on src/_P036_FrameOLED.ino:
```diff
@@ -1598,11 +1603,10 @@ uint8_t display_scroll(int lscrollspeed, int lTaskTimer)
-  if (P036_data == nullptr || !P036_data->isInitialized()) {
-    return;
+  if ((P036_data == nullptr) || !P036_data->isInitialized()) {
+    return (0);
```
You don't have to wrap the return value in braces.
This isn't C.

TD-er commented:
I've tested it and it now indeed seems to work here.

I will now build a test build so others can test it too.
As I said before, this is one of the most used plugins, so it really needs testing.


By the way, see also issue #3127 
The stack usage of this plugin was already high and with this plugin it has doubled.
I will have a look to see if that can be improved as it for sure can cause issues, unless we know for sure it will always be called from the same place with a known stack point at the moment of being called.  (which we don't)

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er commented:
Test build:
- [ESPEasy_ESP32_mega-20200608-89-PR_3050.zip](https://www.dropbox.com/s/cv3s5a1u8rrcqp6/ESPEasy_ESP32_mega-20200608-89-PR_3050.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200608-89-PR_3050.zip](https://www.dropbox.com/s/7xunol3igp81lwq/ESPEasy_ESP82xx_mega-20200608-89-PR_3050.zip?dl=0)

TD-er committed:
Merge branch 'mega' into UweKaditz

uwekaditz committed:
Merge branch 'mega' into UweKaditz

tonhuisman commented:
> The stack usage of this plugin was already high and with this plugin it has doubled.

@uwekaditz : Is this still an issue, after the improvements TD-er did so far?

TD-er commented:
> > The stack usage of this plugin was already high and with this plugin it has doubled.
> 
> @uwekaditz : Is this still an issue, after the improvements TD-er did so far?

Yep, as I have not yet addressed these issues for this plugin.

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
Merge branch 'mega' into UweKaditz

TD-er committed:
[P036_FramedOLED] Reduce stack and heap memory usage

During load, the (large) task data struct was allocated on the heap, regardless whether it was already present in memory when the task was initialized.

During save, the (large) task data struct was allocated on the stack.

Both load and save now check whether there is already an instance present and use that instance.
If not, it creates a temporary instance for loading and saving of the settings.

Also the scrolling variable was not initialized, so it could happen the display would not start to show the pages as it was marked as "still being scrolling"

TD-er committed:
[Web form] Force value in number field within given range

For example the Framed OLED plugin has a default number of lines per page of 0, but the limits of the field only allows 1 .. 4.
This makes it impossible to save the default settings without changing something.

There may be other settings which may have similar issues.

TD-er committed:
[P036_FramedOLED] Uncrustify source code

TD-er commented:
I did have a good look at the code of this PR and made some changes.
Function changes are in this commit: https://github.com/letscontrolit/ESPEasy/pull/3050/commits/6b7ac1e4b053c9550749a29b95388bf6cd4db193

TD-er committed:
[Framed OLED] Move plugin data to separate .h/.cpp

TD-er committed:
[Framed OLED] Clean up code + use temp. object for load/save lines

Loading and saving of display lines must be done using a temporary object and not via the task data object in memory as that may have been changed via write commands.

TD-er commented:
Added a new commit to correct my mistake where I used the in memory task data struct to show data on the web interface and save to the settings.
But the content of those display lines may have changed via commands, so they cannot be used like that.
For showing the web interface, the settings have to be read from the flash.

TD-er commented:
I made a test build:

- [ESPEasy_ESP32_mega-20200721-32-PR_3050.zip](https://www.dropbox.com/s/b1fd16bwsr1sx0t/ESPEasy_ESP32_mega-20200721-32-PR_3050.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200721-32-PR_3050.zip](https://www.dropbox.com/s/1kn9tfe7qv0rgjh/ESPEasy_ESP82xx_mega-20200721-32-PR_3050.zip?dl=0)

uwekaditz commented:
Today I found time to check the new code.
It works well on my devices.

TD-er commented:
@uwekaditz Thanks for testing :)
